be/src/storage/rowset/json_column_writer.cpp

```diff
@@ -0,0 +1,348 @@
+    std::vector<std::string> _flat_paths;
+    std::vector<ColumnPtr> _flat_columns;
+};
+
+FlatJsonColumnWriter::FlatJsonColumnWriter(const ColumnWriterOptions& opts, const TypeInfoPtr& type_info,
+                                           WritableFile* wfile, std::unique_ptr<ScalarColumnWriter> json_writer)
+        : ColumnWriter(std::move(type_info), opts.meta->length(), opts.meta->is_nullable()),
+          _json_column_writer(std::move(json_writer)),
+          _json_meta(opts.meta),
+          _wfile(wfile) {}
+
+Status FlatJsonColumnWriter::append(const Column& column) {
+    return Status::InternalError("json column writer not support append reference interface");
+}
+
+Status FlatJsonColumnWriter::append(const ColumnPtr& column) {
+    size_t ss = column->size();
+    RETURN_IF_ERROR(_json_column_writer->append(column));
+    // keep refs
+    _json_datas.emplace_back(column);
+    for (auto& ptr : _json_datas) {
+        DCHECK_EQ(ss, ptr->size());
+    }
+    return Status::OK();
+}
+
+void FlatJsonColumnWriter::_flat_column(std::vector<ColumnPtr>& json_datas) {
+    DCHECK(_flat_paths.empty());
+    DCHECK(!json_datas.empty());
+
+    size_t total_rows = 0;
+    size_t null_count = 0;
+
+    for (auto& column : json_datas) {
+        DCHECK(!column->is_constant());
+
+        total_rows += column->size();
+        if (column->only_null() || column->empty()) {
+            null_count += column->size();
+            continue;
+        } else if (column->is_nullable()) {
+            auto* nullable_column = down_cast<NullableColumn*>(column.get());
+            null_count += nullable_column->null_count();
+        }
+    }
+
+    // more than half of null
+    if (null_count > total_rows * config::json_flat_null_factor) {
+        VLOG(8) << "flat json, null_count[" << null_count << "], row[" << total_rows
+                << "], null_factor: " << config::json_flat_null_factor;
+        return;
+    }
+
+    // extract common keys
+    std::unordered_map<std::string, uint64_t> hit_maps;
+    for (size_t k = 0; k < json_datas.size(); k++) {
+        size_t row_count = json_datas[k]->size();
+
+        ColumnViewer<TYPE_JSON> viewer(json_datas[k]);
+        for (size_t i = 0; i < row_count; ++i) {
+            if (viewer.is_null(i)) {
+                continue;
+            }
+
+            JsonValue* json = viewer.value(i);
+            auto vslice = json->to_vslice();
+
+            if (vslice.isNone() || vslice.isNull()) {
+                continue;
+            }
+
+            if (!vslice.isObject()) {
+                VLOG(8) << "flat json, row isn't object, can't be flatten";
+                return;
+            }
+
+            std::vector<std::string> keys = arangodb::velocypack::Collection::keys(json->to_vslice());
+            for (auto& sr : keys) {
+                hit_maps[sr]++;
+            }
+        }
+    }
+
+    if (hit_maps.size() <= config::json_flat_internal_column_min_limit) {
```

Comment:
A very stupid question I want to ask,  if a json's keys are parted into two groups, one group is stable and appears a structure, the other group is unstable and used to extending, can we flat the former only?

Reply:
The extract flat column type is JSON now. If only the type is changing (or add/remove a field in nested type), it will still be flattened. If schema fields are add/remov, it won't be flattened. our goal is to extract stable sub-fields.

"In the future, we can maintain query usage statistics in the FE and flatten json based on the user's usage."